File: xbox.cpp
#include <iostream>
#include <fcntl.h>
#include <unistd.h>
#include <linux/joystick.h>

int main() {
    // Open the joystick device
    int joystick_fd = open("/dev/input/js0", O_RDONLY);
    if (joystick_fd == -1) {
        std::cerr << "Could not open joystick device." << std::endl;
        return 1;
    }

    // Get the number of axes and buttons on the joystick
    int num_axes = 0, num_buttons = 0;
    ioctl(joystick_fd, JSIOCGAXES, &num_axes);
    ioctl(joystick_fd, JSIOCGBUTTONS, &num_buttons);
    std::cout << "Joystick connected with " << num_axes << " axes and " << num_buttons << " buttons." << std::endl;

    // Read the joystick events
    struct js_event e;
    while (read(joystick_fd, &e, sizeof(e)) > 0) {
        if (e.type == JS_EVENT_AXIS) {
            // Handle axis events here
            std::cout << "Axis " << static_cast<int>(e.number) << " moved to " << e.value << std::endl;
        } else if (e.type == JS_EVENT_BUTTON) {
            // Handle button events here
            std::cout << "Button " << static_cast<int>(e.number) << " pressed with value " << e.value << std::endl;
        }
    }

    // Close the joystick device
    close(joystick_fd);

    return 0;
}

File: raspberry_pi_version/xbox.py
import os
import struct
import fcntl

device_path = "/dev/input/js0"
joystick_fd = os.open(device_path, os.O_RDONLY)

num_axes = 0
num_buttons = 0
fcntl.ioctl(joystick_fd, 0x80016a11, struct.pack('I', num_axes))
fcntl.ioctl(joystick_fd, 0x80016a12, struct.pack('I', num_buttons))
print("Joystick connected with", num_axes, "axes and", num_buttons, "buttons.")

arr = [0, 0, 0, 0, 0, 0, 0, 0]

while True:
    event = os.read(joystick_fd, 8)
    time, value, event_type, number = struct.unpack("IhBB", event)
    if event_type == 1:
        print("Button", number, "pressed with value", value)
    elif event_type == 2 and number != 1:
        arr[number] = value
        print("Axis", number, "moved to", value)
    print(arr)

os.close(joystick_fd)
